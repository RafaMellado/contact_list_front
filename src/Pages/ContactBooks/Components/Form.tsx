import { FormEvent, useRef } from "react";
import { Button, Card, Form } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import { ErrorBox } from "../../../Components/ErrorBox";
import {
  ContactBook,
  ContactBookRequestBody,
  ContactBookRequestError,
} from "../../../Services/Interfaces/ContactBook";

interface ContactBookFormProps {
  item?: ContactBook;
  translations: string;
  errors?: ContactBookRequestError;
  submitFn: (data: ContactBookRequestBody) => void;
  backFn: () => void;
}

export function ContactBookForm({
  item,
  translations,
  errors,
  submitFn,
  backFn,
}: ContactBookFormProps) {
  const { t } = useTranslation();

  const name = useRef<HTMLInputElement>(null);

  const submit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    submitFn({ name: String(name?.current?.value) });
  };

  return (
    <Card className="form">
      <Card.Body>
        <Form onSubmit={submit}>
          <Form.Group className="mb-3">
            <Form.Label>{t<string>(`${translations}.name`)}</Form.Label>
            <Form.Control
              as="input"
              type="text"
              defaultValue={item?.name}
              placeholder={t<string>(`${translations}.name`)}
              maxLength={20}
              minLength={4}
              required
              ref={name}
            />

            {errors && <ErrorBox errors={errors.errors} field="name" />}
          </Form.Group>

          <div className="d-flex justify-content-between">
            <Button className="w-100" variant="secondary" onClick={backFn}>
              {t<string>(`${translations}.back`)}
            </Button>
            <Button className="w-100 ms-2" variant="primary" type="submit">
              {t<string>(`${translations}.submit`)}
            </Button>
          </div>
        </Form>
      </Card.Body>
    </Card>
  );
}
