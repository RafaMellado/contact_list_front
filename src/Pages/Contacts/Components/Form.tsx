import { FormEvent, useRef } from "react";
import { Button, Card, Form } from "react-bootstrap";
import { useTranslation } from "react-i18next";
import { ErrorBox } from "../../../Components/ErrorBox";
import {
  Contact,
  ContactRequestBody,
  ContactRequestError,
} from "../../../Services/Interfaces/Contact";

export interface ContactFormProps {
  item?: Contact;
  translations: string;
  errors?: ContactRequestError;
  submitFn: (item: Partial<ContactRequestBody>) => void;
  backFn: () => void;
}

export function ContactForm({
  item,
  translations,
  errors,
  submitFn,
  backFn,
}: ContactFormProps) {
  const { t } = useTranslation();

  const givenname = useRef<HTMLInputElement>(null);
  const surname = useRef<HTMLInputElement>(null);
  const email = useRef<HTMLInputElement>(null);
  const phone = useRef<HTMLInputElement>(null);

  const formParams = () => {
    return {
      givenname: String(givenname?.current?.value),
      surname: String(surname?.current?.value),
      email: String(email?.current?.value),
      phone: String(phone?.current?.value),
    };
  };

  const submit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    submitFn(formParams());
  };

  return (
    <Card className="form">
      <Card.Body>
        <Form data-testid="contact-form" onSubmit={submit}>
          <Form.Group className="mb-3">
            <Form.Label>{t<string>(`${translations}.givenname`)}</Form.Label>
            <Form.Control
              as="input"
              type="text"
              defaultValue={item?.givenname}
              placeholder={t<string>(`${translations}.givenname`)}
              maxLength={24}
              minLength={2}
              required
              ref={givenname}
            />
          </Form.Group>

          <Form.Group className="mb-3">
            <Form.Label>{t<string>(`${translations}.surname`)}</Form.Label>
            <Form.Control
              as="input"
              type="text"
              defaultValue={item?.surname}
              placeholder={t<string>(`${translations}.surname`)}
              maxLength={24}
              minLength={2}
              required
              ref={surname}
            />
          </Form.Group>

          <Form.Group className="mb-3">
            <Form.Label>{t<string>(`${translations}.email`)}</Form.Label>
            <Form.Control
              as="input"
              type="email"
              defaultValue={item?.email}
              placeholder={t<string>(`${translations}.email`)}
              required
              ref={email}
            />

            {errors && <ErrorBox errors={errors.errors} field="email" />}
          </Form.Group>

          <Form.Group className="mb-3">
            <Form.Label>{t<string>(`${translations}.phone`)}</Form.Label>
            <Form.Control
              as="input"
              type="tel"
              defaultValue={item?.phone}
              placeholder={t<string>(`${translations}.phone`)}
              maxLength={12}
              required
              ref={phone}
            />

            {errors && <ErrorBox errors={errors.errors} field="phone" />}
          </Form.Group>

          <div className="d-flex justify-content-between">
            <Button
              data-testid="contact-form-back-btn"
              className="w-100"
              variant="secondary"
              onClick={() => backFn()}
            >
              {t<string>(`${translations}.backBtn`)}
            </Button>

            <Button className="w-100 ms-3" variant="primary" type="submit">
              {t<string>(`${translations}.submit`)}
            </Button>
          </div>
        </Form>
      </Card.Body>
    </Card>
  );
}
