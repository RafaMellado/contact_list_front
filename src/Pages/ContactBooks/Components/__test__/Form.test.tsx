import { ContactBookForm, ContactBookFormProps } from "../Form";
import { fireEvent, render } from "@testing-library/react";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

describe(`<${ContactBookForm.name} />`, () => {
  const submitFn = jest.fn();
  const backFn = jest.fn();

  const defaultProps: ContactBookFormProps = {
    translations: "test",
    submitFn: submitFn,
    backFn: backFn,
    item: {
      id: 1,
      name: "test",
      user_id: 1,
    },
  };

  const factoryComponent = (props: ContactBookFormProps = defaultProps) =>
    render(<ContactBookForm {...props} />);

  test("renders correctly the component", () => {
    const component = factoryComponent();

    expect(component.asFragment()).toMatchSnapshot();
  });

  test("submit calls submitFn correctly", () => {
    const component = factoryComponent();

    fireEvent.submit(component.getByTestId("contact-book-form"));

    expect(submitFn).toHaveBeenCalledWith({ name: "test" });
  });

  test("click on backBtn redirects correctly", () => {
    const component = factoryComponent();

    fireEvent.click(component.getByTestId("contact-book-back-btn"));

    expect(backFn).toHaveBeenCalled();
  });

  test("show errors correctly", () => {
    const component = factoryComponent({
      ...defaultProps,
      errors: { errors: { name: [{ error: "test", value: "test value" }] } },
    });

    expect(component.asFragment()).toMatchSnapshot("errors");
  });
});
