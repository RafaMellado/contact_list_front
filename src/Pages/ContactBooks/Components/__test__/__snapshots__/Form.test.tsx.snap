// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactBookForm /> renders correctly the component 1`] = `
<DocumentFragment>
  <div
    class="form card"
  >
    <div
      class="card-body"
    >
      <form
        class=""
        data-testid="contact-book-form"
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            test.name
          </label>
          <input
            class="form-control"
            maxlength="20"
            minlength="4"
            placeholder="test.name"
            required=""
            type="text"
            value="test"
          />
        </div>
        <div
          class="d-flex justify-content-between"
        >
          <button
            class="w-100 btn btn-secondary"
            data-testid="contact-book-back-btn"
            type="button"
          >
            test.back
          </button>
          <button
            class="w-100 ms-2 btn btn-primary"
            type="submit"
          >
            test.submit
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ContactBookForm /> show errors correctly: errors 1`] = `
<DocumentFragment>
  <div
    class="form card"
  >
    <div
      class="card-body"
    >
      <form
        class=""
        data-testid="contact-book-form"
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            test.name
          </label>
          <input
            class="form-control"
            maxlength="20"
            minlength="4"
            placeholder="test.name"
            required=""
            type="text"
            value="test"
          />
          <div
            class="text-danger mt-2 mb-0"
            data-testid="error-box-name"
          >
            <span>
              errors.test
            </span>
          </div>
        </div>
        <div
          class="d-flex justify-content-between"
        >
          <button
            class="w-100 btn btn-secondary"
            data-testid="contact-book-back-btn"
            type="button"
          >
            test.back
          </button>
          <button
            class="w-100 ms-2 btn btn-primary"
            type="submit"
          >
            test.submit
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
